### DISCUSSION TITLE
maven-based target platform definition: option to generate new manifest even if one already exists?

### DISCUSSION THREAD

**Discussion by sratz** [eclipse-tycho/tycho#1519] [Ideas]
The `missingManifest="generate"` option in maven based target definition allows me to generate an OSGi bundle, if it is not one already.

However, we have a use-case where the artifact already is an OSGi bundle, but we need to modify it for consumption.

What do you think about providing a new option like `forceGenerateManifest="true"` to allow this?
(Alternatively a 4th alternative for the existing  `missingManifest` option, although that would not really fit, since the manifest isn't actually missing).

This would allow us to fully migrate from https://github.com/reficio/p2-maven-plugin to Tycho.

**Comment by laeubi** [comment]
Can you explain why you need a different manifest? If there are any bugs it might be better to ask the provider to fix them?

**Reply by sratz** [reply]
This has to do with a legacy custom non-Eclipse/Tycho-based build system that produces inconsistent manifests for SNAPSHOT builds:
Concretely, for SNAPSHOT maven exports, the generated MANIFEST.MF is missing the .qualifier in the version:

maven version:
`1.0.0-SNAPSHOT (1.0.0-20221012.124900-5)`

but MANIFEST.MF:
`Bundle-Version: 1.0.0`
instead of
`Bundle-Version: 1.0.0.v20221012124900`

The p2-maven-plugin
1) takes care of such inconsistencies automatically using some utilities: https://github.com/reficio/p2-maven-plugin/blob/master/src/main/java/org/reficio/p2/utils/Utils.java
2) Allows me to provide BND instructions to override existing manifest

I would like to migrate to Tycho and would like to consume those snapshot bundles with Tycho, but fix the manifest to contain qualified versions, such that newer snapshot builds get higher version numbers and things can be updated in the final product built by the Tycho build.

**Reply by sratz** [reply]
Implementation-wise in `org.eclipse.tycho.p2.resolver.MavenTargetDefinitionContent` such a 4th option would be rather simple, but would offer a powerful tool to work around broken `bundle` providers.

**Reply by laeubi** [reply]
The target support is not really meant to fix broken manifests, so I'm not very convinced we should add those, also there is no requirement that the maven version matches the bundle version...

**Reply by sratz** [reply]
> The target support is not really meant to fix broken manifests

I agree, but since the infrastructure to generate them is there now, the fixing part can trivially be implemented without much effort.

> here is no requirement that the maven version matches the bundle version

Maybe there are other scenarios, where one might want to modify/add BND instructions.

Such an option would be quite helpful. It should of course be documented accordingly that it's preferred to not touch existing bundles and considered an advanced option.

Otherwise, I am kind of stuck with the old legacy https://github.com/reficio/p2-maven-plugin which does not appear to be actively developed anymore.

**Reply by laeubi** [reply]
> > The target support is not really meant to fix broken manifests
> 
> I agree, but since the infrastructure to generate them is there now, the fixing part can trivially be implemented without much effort.

First, as you have already noticed, its not quite clear how this could be implemented in a way that seems not "odd" (e.g. another option for "missing manifest" that replaces a manifest sound really odd). Second one has to test and support such option.

Next the "generate" option is more meant for intermediate usage than as a replacement for any tooling constraints, managing many items with specialized BND instructions will result in large target files that are hard to manage and understand.

>> here is no requirement that the maven version matches the bundle version
> 
> Maybe there are other scenarios, where one might want to modify/add BND instructions.

A manifest is not necessarily created out of BND instructions, and it might then be more valuable to have a real BND step doing what is desired https://bnd.bndtools.org/chapters/390-wrapping.html and later deply the result unter a new dedicated GAV

> Such an option would be quite helpful. It should of course be documented accordingly that it's preferred to not touch existing bundles and considered an advanced option.
> Otherwise, I am kind of stuck with the old legacy https://github.com/reficio/p2-maven-plugin which does not appear to be actively developed anymore.

I would more consider it a workaround :-) Especially if you are talking about snapshot builds, it seems it will more help in not fixing the build-issue than solving a real problem. Also such an option would mean that always all manifests of a location are recreated, what will further limit the usefulness, because then you most likely will require one location and one bnd instruction for every single artifact and can't use transitive dependencies to be included automatically.

**Reply by akurtakov** [reply]
IMHO Tycho should not add such option. That would be extremely error-prone option.

**Reply by mickaelistria** [reply]
I agree with Alex. What's suggested here is extremely likely to create "unofficial" forked artifacts, that are then published in p2 repo, mappend to Maven coordinates of an artifact that is in practice different... Implementing this would IMO be more than error-prone, but even harmful to the ecosystem.

The proper solution to "we have a use-case where the artifact already is an OSGi bundle, but we need to modify it for consumption" is to fix that particular case with a particular solution (fixing the artifact, or creating another artifact exporting same packages but with some fix and requesting the fixed one in clients).
